Add maxout op to tf.contrib.layers

tiagofrepereira2012 opened the pull request:
Hey,

I just wanted to activate this PR #5528 

- Ported maxout layer from !5528
- Improved documentation
- Created test units

Could someone please review this one?

Thanks

tensorflow-jenkins commented:
Can one of the admins verify this patch?

tiagofrepereira2012 commented:
I signed it!

alextp commented on tensorflow/contrib/layers/python/layers/layers.py:
```diff
@@ -104,44 +106,44 @@ def avg_pool2d(inputs,
                data_format=DATA_FORMAT_NHWC,
                outputs_collections=None,
                scope=None):
-  """Adds a 2D average pooling op.
```
These large format-only changes make it really difficult for me to review this change. Can you revert all formatting changes and send only the functional ones?

tiagofrepereira2012 replied:
Changes made.
Could you please check it out?

alextp commented on tensorflow/contrib/layers/python/layers/layers.py:
```diff
@@ -2163,45 +2175,93 @@ def max_pool3d(inputs,
+
+
+@add_arg_scope
+def maxout(inputs,
```
Can you implement maxout in the style of tf.layers (with a class) as opposed to the style of tf.contrib.layers? This way there's a path to merging it in tf core.

tiagofrepereira2012 replied:
Sure, shall I create a specific file for that ?

Mazecreator commented:
This message was created automatically by mail delivery software.

A message that you sent could not be delivered to one or more of its
recipients. This is a temporary error. The following address(es) deferred:

  mazecreator@gmail.com
    Domain mazecreator.com has exceeded the max emails per hour (25/25 (100%)) allowed.  Message will be reattempted later

------- This is a copy of the message, including all the headers. ------
Received: from github-smtp2-ext3.iad.github.net ([192.30.252.194]:51275 helo=github-smtp2b-ext-cp1-prd.iad.github.net)
	by server2.lowesthostingrates.com with esmtps (TLSv1.2:ECDHE-RSA-AES256-GCM-SHA384:256)
	(Exim 4.89)
	(envelope-from <noreply@github.com>)
	id 1dcv2h-0005IT-Mo
	for mazecreator@mazecreator.com; Wed, 02 Aug 2017 09:55:05 -0500
Date: Wed, 02 Aug 2017 08:05:15 -0700
DKIM-Signature: v=1; a=rsa-sha256; c=relaxed/relaxed; d=github.com;
	s=pf2014; t=1501686315;
	bh=FfdDhXGZL8lXJKdnq6ZnjpeEbvP7tAM16+QYP2tsirc=;
	h=From:Reply-To:To:Cc:In-Reply-To:References:Subject:List-ID:
	 List-Archive:List-Post:List-Unsubscribe:From;
	b=RsYfkylWFieyQ0qlV1smmKuhvHS4Knf3XA6wkeLVcp337RJroqZ3vyrglB+oYFAHv
	 aEzEkRosppZ2m7G8s/E12M+CfpHWDsU8bgUuqdL0fVI9yVbmwceKfkBQ9yf1Mz54kY
	 YrzrrKRREoQjacxaevf3hf74PbylUBh71cwyt2/0=
From: Tiago Freitas Pereira <notifications@github.com>
Reply-To: tensorflow/tensorflow <reply@reply.github.com>
To: tensorflow/tensorflow <tensorflow@noreply.github.com>
Cc: Subscribed <subscribed@noreply.github.com>
Message-ID: <tensorflow/tensorflow/pull/11824/issue_event/1189721598@github.com>
In-Reply-To: <tensorflow/tensorflow/pull/11824@github.com>
References: <tensorflow/tensorflow/pull/11824@github.com>
Subject: Re: [tensorflow/tensorflow] Add maxout op to tf.contrib.layers
 (#11824)
Mime-Version: 1.0
Content-Type: multipart/alternative;
 boundary="--==_mimepart_5981ea2b3fc4a_22323fd4ec969c303779af";
 charset=UTF-8
Content-Transfer-Encoding: 7bit
Precedence: list
X-GitHub-Sender: tiagofrepereira2012
X-GitHub-Recipient: Mazecreator
X-GitHub-Reason: subscribed
List-ID: tensorflow/tensorflow <tensorflow.tensorflow.github.com>
List-Archive: https://github.com/tensorflow/tensorflow
List-Post: <mailto:reply@reply.github.com>
List-Unsubscribe: <mailto:unsub+0118f3a0e174358a31fa4c19e53ac716c2ad3eaf826f692992cf000000011599ac2b92a169ce0eaba4bd@reply.github.com>,
 <https://github.com/notifications/unsubscribe/ARjzoKh-6QcONiZ1rVIU1XiwEhSJAptkks5sUJArgaJpZM4Olq41>
X-Auto-Response-Suppress: All
X-GitHub-Recipient-Address: mazecreator@mazecreator.com
X-Spam-Status: No, score=
X-Spam-Score:
X-Spam-Bar:
X-Ham-Report:
X-Spam-Flag: NO


----==_mimepart_5981ea2b3fc4a_22323fd4ec969c303779af
Content-Type: text/plain;
 charset=UTF-8
Content-Transfer-Encoding: 7bit

Closed #11824.

-- 
You are receiving this because you are subscribed to this thread.
Reply to this email directly or view it on GitHub:
https://github.com/tensorflow/tensorflow/pull/11824#event-1189721598
----==_mimepart_5981ea2b3fc4a_22323fd4ec969c303779af
Content-Type: text/html;
 charset=UTF-8
Content-Transfer-Encoding: 7bit

<p>Closed <a href="https://github.com/tensorflow/tensorflow/pull/11824" class="issue-link js-issue-link" data-url="https://github.com/tensorflow/tensorflow/issues/11824" data-id="246129853" data-error-text="Failed to load issue title" data-permission-text="Issue title is private">#11824</a>.</p>

<p style="font-size:small;-webkit-text-size-adjust:none;color:#666;">&mdash;<br />You are receiving this because you are subscribed to this thread.<br />Reply to this email directly, <a href="https://github.com/tensorflow/tensorflow/pull/11824#event-1189721598">view it on GitHub</a>, or <a href="https://github.com/notifications/unsubscribe-auth/ARjzoPsktuGHrbDjOJG3i3_snUsihn0aks5sUJArgaJpZM4Olq41">mute the thread</a>.<img alt="" height="1" src="https://github.com/notifications/beacon/ARjzoHbCZ4OOKUu5BQHX7d8DQ0vrrIC8ks5sUJArgaJpZM4Olq41.gif" width="1" /></p>
<div itemscope itemtype="http://schema.org/EmailMessage">
<div itemprop="action" itemscope itemtype="http://schema.org/ViewAction">
  <link itemprop="url" href="https://github.com/tensorflow/tensorflow/pull/11824#event-1189721598"></link>
  <meta itemprop="name" content="View Pull Request"></meta>
</div>
<meta itemprop="description" content="View this Pull Request on GitHub"></meta>
</div>

<script type="application/json" data-scope="inboxmarkup">{"api_version":"1.0","publisher":{"api_key":"05dde50f1d1a384dd78767c55493e4bb","name":"GitHub"},"entity":{"external_key":"github/tensorflow/tensorflow","title":"tensorflow/tensorflow","subtitle":"GitHub repository","main_image_url":"https://cloud.githubusercontent.com/assets/143418/17495839/a5054eac-5d88-11e6-95fc-7290892c7bb5.png","avatar_image_url":"https://cloud.githubusercontent.com/assets/143418/15842166/7c72db34-2c0b-11e6-9aed-b52498112777.png","action":{"name":"Open in GitHub","url":"https://github.com/tensorflow/tensorflow"}},"updates":{"snippets":[{"icon":"DESCRIPTION","message":"Closed #11824."}],"action":{"name":"View Pull Request","url":"https://github.com/tensorflow/tensorflow/pull/11824#event-1189721598"}}}</script>
----==_mimepart_5981ea2b3fc4a_22323fd4ec969c303779af--

alextp commented:
Up to you, but I think smaller files are better for many reasons

On Wed, Aug 2, 2017 at 8:05 AM, Tiago Freitas Pereira <
notifications@github.com> wrote:

> Closed #11824 <https://github.com/tensorflow/tensorflow/pull/11824>.
>
> —
> You are receiving this because you were assigned.
> Reply to this email directly, view it on GitHub
> <https://github.com/tensorflow/tensorflow/pull/11824#event-1189721598>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAATxQTKazBTIO-xvMfS5bv0bHNvLRstks5sUJA9gaJpZM4Olq41>
> .
>



-- 
 - Alex

tiagofrepereira2012 committed:
Implemented maxout

Removed None test from the maxout clas

Fixed test units

Fixed import and lint the code

Moved entry point to the contrib module

pylint the code

pylint

Organizing build

Trying to figure the buildifier issue

tensorflow-jenkins commented:
Can one of the admins verify this patch?

tiagofrepereira2012 commented:
So, the changes were made.
Please let me know if there's anything (I will be off for three weeks).  
In the future I will put in place this feature https://arxiv.org/abs/1511.02683

alextp commented on tensorflow/python/layers/maxout.py:
```diff
@@ -0,0 +1,111 @@
+    def call(self, inputs, training=False):
+        inputs = ops.convert_to_tensor(inputs)
+        shape = inputs.get_shape().as_list()
+        if self.axis is None:
```
This will only be None if the user explicitly passes None, which is a little weird. How about making the default value in the constructor None so -1 only appears in a single place in code?

tiagofrepereira2012 replied:
I'm back.

Yes, you are right. I'll patch that.

alextp commented on tensorflow/python/layers/maxout_test.py:
```diff
@@ -0,0 +1,62 @@
+from __future__ import division
+from __future__ import print_function
+
+from tensorflow.python.layers.maxout import maxout
```
Do

from tensorflow.python.layers import maxout

as this is more in keeping with the style of the code.

tiagofrepereira2012 replied:
sure

alextp commented on tensorflow/python/layers/layers.py:
```diff
@@ -31,6 +31,7 @@
 @@average_pooling3d
 @@max_pooling3d
 @@batch_normalization
+@@maxout
```
Instead of exposing the symbol here, how about exposing it from contrib.layers (leaving the code where it is)? This way we can work out if any APIs need to change before committing to backward compatibility forever.

tiagofrepereira2012 replied:
Hmmm, I don't understand why this symbol is not suitable here

alextp replied:
Adding the symbol here will add tf.layers.maxout (but not tf.layers.Maxout, the object-based version). In tensorflow, since 1.0, we are promising full backward compatibility on all APIs which are not in tf.contrib. Since maxout was just added I don't think it deserves full backward compatibility promises yet. So I prefer that the symbol is added to tf.contrib.layers.{Maxout,maxout}

tiagofrepereira2012 replied:
Ok, understood.
Shall I keep the code in the same place and just import the symbol in the contrib or I should move everything?

alextp replied:
Yes, the code is in the right place.

tiagofrepereira2012 replied:
Done, but I still don't know how to make the sanity checks green.
The unit tests and pylint works in my workstation

Shall I squash my commits everything before merge?

alextp commented:
Rebasing is a good idea.

Your errors come from pylint and buildifier. I'll paste them here:

FAIL: buildifier found errors and/or warnings in above BUILD files.
buildifier suggested the following changes:
3699d3698
<         "layers/maxout.py",
3701a3701
>         "layers/maxout.py",
3793d3792
< 

This means your source files need to be sorted in the BUILD rule.

Then the pylint errors are

pylint took 116 s


FAIL: Found 39 non-whitelited pylint errors:
tensorflow/python/layers/maxout_test.py:37: [W0311(bad-indentation), ] Bad indentation. Found 4 spaces, expected 2

tensorflow/python/layers/maxout_test.py:38: [W0311(bad-indentation), ] Bad indentation. Found 8 spaces, expected 4

tensorflow/python/layers/maxout_test.py:39: [W0311(bad-indentation), ] Bad indentation. Found 8 spaces, expected 4

tensorflow/python/layers/maxout_test.py:40: [W0311(bad-indentation), ] Bad indentation. Found 8 spaces, expected 4

tensorflow/python/layers/maxout_test.py:42: [W0311(bad-indentation), ] Bad indentation. Found 4 spaces, expected 2

tensorflow/python/layers/maxout_test.py:43: [W0311(bad-indentation), ] Bad indentation. Found 8 spaces, expected 4

tensorflow/python/layers/maxout_test.py:44: [W0311(bad-indentation), ] Bad indentation. Found 8 spaces, expected 4

tensorflow/python/layers/maxout_test.py:45: [W0311(bad-indentation), ] Bad indentation. Found 8 spaces, expected 4

tensorflow/python/layers/maxout_test.py:46: [W0311(bad-indentation), ] Bad indentation. Found 8 spaces, expected 4

tensorflow/python/layers/maxout_test.py:48: [W0311(bad-indentation), ] Bad indentation. Found 4 spaces, expected 2

tensorflow/python/layers/maxout_test.py:49: [W0311(bad-indentation), ] Bad indentation. Found 8 spaces, expected 4

tensorflow/python/layers/maxout_test.py:50: [W0311(bad-indentation), ] Bad indentation. Found 8 spaces, expected 4

tensorflow/python/layers/maxout_test.py:51: [W0311(bad-indentation), ] Bad indentation. Found 8 spaces, expected 4

tensorflow/python/layers/maxout_test.py:52: [W0311(bad-indentation), ] Bad indentation. Found 8 spaces, expected 4

tensorflow/python/layers/maxout_test.py:54: [W0311(bad-indentation), ] Bad indentation. Found 4 spaces, expected 2

tensorflow/python/layers/maxout_test.py:55: [W0311(bad-indentation), ] Bad indentation. Found 8 spaces, expected 4

tensorflow/python/layers/maxout_test.py:56: [W0311(bad-indentation), ] Bad indentation. Found 8 spaces, expected 4

tensorflow/python/layers/maxout_test.py:57: [W0311(bad-indentation), ] Bad indentation. Found 8 spaces, expected 4

tensorflow/python/layers/maxout_test.py:58: [W0311(bad-indentation), ] Bad indentation. Found 12 spaces, expected 6

tensorflow/python/layers/maxout.py:31: [W0311(bad-indentation), ] Bad indentation. Found 4 spaces, expected 2

tensorflow/python/layers/maxout.py:55: [W0311(bad-indentation), ] Bad indentation. Found 4 spaces, expected 2

tensorflow/python/layers/maxout.py:59: [W0311(bad-indentation), ] Bad indentation. Found 4 spaces, expected 2

tensorflow/python/layers/maxout.py:83: [W0311(bad-indentation), ] Bad indentation. Found 4 spaces, expected 2

tensorflow/python/layers/maxout.py:88: [W0311(bad-indentation), ] Bad indentation. Found 8 spaces, expected 4

tensorflow/python/layers/maxout.py:90: [W0311(bad-indentation), ] Bad indentation. Found 8 spaces, expected 4

tensorflow/python/layers/maxout.py:91: [W0311(bad-indentation), ] Bad indentation. Found 8 spaces, expected 4

tensorflow/python/layers/maxout.py:93: [W0311(bad-indentation), ] Bad indentation. Found 4 spaces, expected 2

tensorflow/python/layers/maxout.py:94: [W0311(bad-indentation), ] Bad indentation. Found 8 spaces, expected 4

tensorflow/python/layers/maxout.py:95: [W0311(bad-indentation), ] Bad indentation. Found 8 spaces, expected 4

tensorflow/python/layers/maxout.py:96: [W0311(bad-indentation), ] Bad indentation. Found 8 spaces, expected 4

tensorflow/python/layers/maxout.py:97: [W0311(bad-indentation), ] Bad indentation. Found 8 spaces, expected 4

tensorflow/python/layers/maxout.py:98: [W0311(bad-indentation), ] Bad indentation. Found 12 spaces, expected 6

tensorflow/python/layers/maxout.py:101: [W0311(bad-indentation), ] Bad indentation. Found 8 spaces, expected 4

tensorflow/python/layers/maxout.py:102: [W0311(bad-indentation), ] Bad indentation. Found 8 spaces, expected 4

tensorflow/python/layers/maxout.py:105: [W0311(bad-indentation), ] Bad indentation. Found 8 spaces, expected 4

tensorflow/python/layers/maxout.py:106: [W0311(bad-indentation), ] Bad indentation. Found 12 spaces, expected 6

tensorflow/python/layers/maxout.py:107: [W0311(bad-indentation), ] Bad indentation. Found 16 spaces, expected 8

tensorflow/python/layers/maxout.py:108: [W0311(bad-indentation), ] Bad indentation. Found 8 spaces, expected 4

tensorflow/python/layers/maxout.py:110: [W0311(bad-indentation), ] Bad indentation. Found 8 spaces, expected 4

to fix these use 2-space indentation instead of 4-space indentation

tiagofrepereira2012 commented:
Thanks for the prints, I managed to solve the issues with pylint, but the sanity checks are still not green and the log is not very clear (http://ci.tensorflow.org/job/tensorflow-pull-requests-sanity/6766/consoleFull).

Can I have some help?

Thanks

alextp commented:
Add a blank line above line 3794 of tensorflow/python/BUILD to make buildifier happy.

For the py_test error I have no idea what it means, and the instruction to fix seems wrong to me. Let's make the other change and rerun the tests.

tiagofrepereira2012 commented:
All green :-)

Thanks for the help

martinwicke commented:
Jenkins, test this please.

tiagofrepereira2012 commented:
I observed that the failures in the tests are not related with my contribution. 

Shall I do something?

Cheers

alextp commented:
Jenkins, test this please.

martinwicke commented:
Jenkins, test this please.

There were issues with the executors.

tiagofrepereira2012 commented:
Any news from this one?

Can I help with something?

alextp commented:
Jenkins, test this please.
